timeline_sync/models.py

```diff
@@ -0,0 +1,230 @@
+
+    def update_from_json(self, pin_json):
+        self.time = valid_time(parse_time(pin_json['time']))
+        self.duration = pin_json.get('duration')
+        self.create_notification = TimelineNotification.from_json(pin_json.get('createNotification'), optional_time=True)
+        self.update_notification = TimelineNotification.from_json(pin_json.get('updateNotification'))
+        self.layout = TimelineLayout.from_json(pin_json['layout'])
+        self.reminders = TimelineReminder.from_json(pin_json.get('reminders'))
+        self.actions = TimelineAction.from_json(pin_json.get('actions'))
+        self.update_time = datetime.datetime.utcnow()
+
+    def to_json(self):
+        result = {
+            "time": time_to_str(self.time),
+            "layout": self.layout.to_json(),
+            "guid": self.guid,
+            "dataSource": self.data_source,
+            "source": self.source,
+            "createTime": time_to_str(self.create_time),
+            "updateTime": time_to_str(self.update_time),
+            "topicKeys": self.topic_keys
+        }
+
+        if self.duration is not None:
+            result["duration"] = self.duration
+        if self.create_notification is not None:
+            result["createNotification"] = self.create_notification.to_json()
+        if self.update_notification is not None:
+            result["updateNotification"] = self.update_notification.to_json()
+        if self.reminders is not None and len(self.reminders) > 0:
+            result["reminders"] = [reminder.to_json() for reminder in self.reminders]
+        if self.actions is not None and len(self.actions) > 0:
+            result["actions"] = [action.to_json() for action in self.actions]
+
+        return result
+
+
+db.Index('timeline_pin_appuuid_uid_pinid_index', TimelinePin.app_uuid, TimelinePin.user_id, TimelinePin.id, unique=True)
+
+
+class TimelineNotification(db.Model):
+    __tablename__ = "timeline_notifications"
+    id = db.Column(db.Integer, primary_key=True)
+    time = db.Column(db.DateTime)
+
+    layout_id = db.Column(db.Integer, db.ForeignKey('timeline_layouts.id'), nullable=False)
+    layout = db.relationship('TimelineLayout', lazy=False, uselist=False, single_parent=True, cascade="all, delete-orphan")
+
+    @classmethod
+    def from_json(cls, notification_json, optional_time=False):
+        if notification_json is None:
+            return notification_json
+        try:
+            time = valid_time(parse_time(notification_json['time']))
+        except KeyError as e:
+            if optional_time:
+                time = None
+            else:
+                raise e
+        return cls(
+            time=time,
+            layout=TimelineLayout.from_json(notification_json['layout'])
+        )
+
+    def to_json(self):
+        result = {"layout": self.layout.to_json()}
+        if self.time:
+            result["time"] = time_to_str(self.time)
+        return result
+
+
+class TimelineLayout(db.Model):
```

Comment:
What is the rationale for pulling these into their own table instead of having inline jsonb fields on the relevant tables?